component/loki/source/podlogs/controller.go

```diff
@@ -0,0 +1,196 @@
+			return fmt.Errorf("unable to register scheme: %w", err)
+		}
+	}
+
+	cli, err := client.New(cfg, client.Options{Scheme: scheme})
+	if err != nil {
+		return err
+	}
+
+	cache, err := cache.New(cfg, cache.Options{Scheme: scheme})
+	if err != nil {
+		return err
+	}
+
+	delegateCli, err := client.NewDelegatingClient(client.NewDelegatingClientInput{
+		CacheReader: cache,
+		Client:      cli,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Update the stored informers and client and schedule a reload.
+	ctrl.mut.Lock()
+	ctrl.informers = cache
+	ctrl.client = delegateCli
+	ctrl.mut.Unlock()
+
+	select {
+	case ctrl.reloadCh <- struct{}{}:
+	default:
+		// Reload is already scheduled
+	}
+	return nil
+}
+
+// Run the controller.
+func (ctrl *controller) Run(ctx context.Context) error {
+	var (
+		cancel    context.CancelFunc
+		informers cache.Informers
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ctrl.reloadCh:
+			ctrl.mut.RLock()
+			var (
+				newInformers = ctrl.informers
+				newClient    = ctrl.client
+			)
+			ctrl.mut.RUnlock()
+
+			// Stop old informers.
+			if informers != nil {
+				cancel()
+			}
+
+			informerContext, informerCancel := context.WithCancel(ctx)
+
+			go func() {
+				if err := ctrl.run(informerContext, newInformers, newClient); err != nil {
+					level.Error(ctrl.log).Log("msg", "failed to run controller", "err", err)
+				}
+			}()
+
+			cancel = informerCancel
+			informers = newInformers
+		}
+	}
+}
+
+func (ctrl *controller) run(ctx context.Context, informers cache.Informers, client client.Client) error {
+	level.Info(ctrl.log).Log("msg", "starting controller")
+	defer level.Info(ctrl.log).Log("msg", "controller exiting")
+
+	go func() {
+		err := informers.Start(ctx)
+		if err != nil && ctx.Err() != nil {
+			level.Error(ctrl.log).Log("msg", "failed to start informers", "err", err)
+		}
+	}()
+
+	if !informers.WaitForCacheSync(ctx) {
+		return fmt.Errorf("informer caches failed to sync")
+	}
+
+	if err := ctrl.configureInformers(ctx, informers); err != nil {
+		return fmt.Errorf("failed to configure informers: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ctrl.reconcileCh:
+			if err := ctrl.reconciler.Reconcile(ctx, client); err != nil {
+				level.Error(ctrl.log).Log("msg", "reconcile failed", "err", err)
+			}
+		}
+	}
+}
+
+// configureInformers starts the informers used by this controller to perform reconciles.
+func (ctrl *controller) configureInformers(ctx context.Context, informers cache.Informers) error {
+	// We want to re-reconcile the set of PodLogs whenever namespaces, pods, or
+	// PodLogs changes. Reconciling on namespaces and pods is important so that
+	// we can reevaluate selectors defined in PodLogs.
+	types := []client.Object{
+		&corev1.Namespace{},
+		&corev1.Pod{},
+		&monitoringv1alpha2.PodLogs{},
+	}
+	for _, ty := range types {
+		informer, err := informers.GetInformer(ctx, ty)
+		if err != nil {
+			return err
```

Comment:
If we error early do we need to do anything with all the `onChangeEventHandler` remember the danging events from .net days.